components/plurals/src/provider.rs

```diff
@@ -319,3 +319,164 @@ mod ranges {
+        }
+
+        if op.is_exactly_one() {
+            if let Some(value) = self
+                .specials
+                .iter()
+                .filter_map(|ule| {
+                    (ule.0 == PluralElementsKeysV1::ExplicitZero.to_unaligned()).then_some(&ule.1)
+                })
+                .next()
+            {
+                return value;
+            }
+        }
+
+        let category = match category {
+            PluralCategory::Zero => PluralElementsKeysV1::Zero,
+            PluralCategory::One => PluralElementsKeysV1::One,
+            PluralCategory::Two => PluralElementsKeysV1::Two,
+            PluralCategory::Few => PluralElementsKeysV1::Few,
+            PluralCategory::Many => PluralElementsKeysV1::Many,
+            PluralCategory::Other => return &self.other,
+        }
+        .to_unaligned();
+        self.specials
+            .iter()
+            .filter_map(|ule| (ule.0 == category).then_some(&ule.1))
+            .next()
+            .unwrap_or(&*self.other)
+    }
+}
+
+#[cfg(feature = "datagen")]
+impl<'a, 'b> From<PluralElements<'b, str>> for PluralElementsV1<'a> {
+    fn from(value: PluralElements<'b, str>) -> Self {
+        Self {
+            specials: (&[
+                value
+                    .zero
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::Zero, s.into())),
+                value
+                    .one
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::One, s.into())),
+                value
+                    .two
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::Two, s.into())),
+                value
+                    .few
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::Few, s.into())),
+                value
+                    .many
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::Many, s.into())),
+                value
+                    .explicit_zero
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::ExplicitZero, s.into())),
+                value
+                    .explicit_one
+                    .filter(|p| *p != value.other)
+                    .map(|s| PluralElementsFieldV1(PluralElementsKeysV1::ExplicitOne, s.into())),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>())
+                .into(),
+            other: value.other.to_owned().into(),
+        }
+    }
+}
+
+#[test]
+fn plural_elements_niche() {
+    assert_eq!(core::mem::size_of::<PluralElementsV1>(), 48);
+    assert_eq!(core::mem::size_of::<Option<PluralElementsV1>>(), 48);
+}
```

Comment:
Nit: 48 B is the size of these things in finely sliced data markers. It would be nice if it could be 24 B or 32 B at the most.

Fine to land this PR first and refactor the representation later (before currency graduation).

Reply:
This is a preexisting struct, I'm just moving it around. I've already optimised this a lot, and will leave it like this for now.